### PR TITLE
iOS and MacOS Native support

### PR DESCRIPTION
**Pull request by DevSrSouza**
Known issues with Compose

- ~~iOS Build fail when `androidMain` was resource files, apparently when building for iOS, the plugin is also using resources from other targets (in this case, `androidMain`).~~ (https://github.com/JetBrains/compose-jb/pull/1984)
- BackHandler fail at build time when BackHandler was a default value.
- LazyColumn fails to start applications

TODO:
- [ ] Support iOS back handler (take a look at the Wear OS Swipe)
- [ ] Revisit AtomicInt32 or maybe replace with [compose runtime AtomicInt](https://android-review.googlesource.com/c/platform/frameworks/support/+/1760172).
- [ ] ThreadSafeMap
- [ ] Disable ios and mac on non darwin systems (Windows/Linux)
- [ ] Setup a CI for publishing to Maven Central on version release. Allowing setup 2 workflows, one special for macOS and iOS native publishing.

https://user-images.githubusercontent.com/29736164/160265768-16ecb936-d458-48ec-a828-de8f3369a424.mov

### CODE REVIEW

**Comment by terrakok** [comment]
LazyColumn already works :)

**Comment by DevSrSouza** [comment]
I Will be working on the PR this weekend.
Hopefully next week we can have a experimental artifact from Voyager running on Native / iOS

File: voyager-core/src/nativeMain/kotlin/cafe.adriel.voyager.core/concurrent/AtomicInt32.native.kt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package cafe.adriel.voyager.core.concurrent
+
+public actual class AtomicInt32 actual constructor(initialValue: Int) {
+    private val delegate = kotlin.native.concurrent.AtomicInt(initialValue)
+    public actual fun getAndIncrement(): Int {
+        return delegate.addAndGet(1) // FIXME: should use increment() ?
+    }
+}
```
Comment (Syer10) [inline voyager-core/src/nativeMain/kotlin/cafe.adriel.voyager.core/concurrent/AtomicInt32.native.kt:8]:
```suggestion
package cafe.adriel.voyager.core.concurrent

import kotlinx.atomicfu.atomic

public actual class AtomicInt32 actual constructor(initialValue: Int) {
    private val delegate = atomic(initialValue)
    public actual fun getAndIncrement(): Int {
        return delegate.getAndIncrement()
    }
}

```

File: voyager-core/src/nativeMain/kotlin/cafe.adriel.voyager.core/concurrent/ThreadSafeMap.native.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package cafe.adriel.voyager.core.concurrent
+
+// FIXME
+public actual class ThreadSafeMap<K, V>(
+    private val delegate: MutableMap<K, V>
+) : MutableMap<K, V> by delegate {
+    public actual constructor() : this(delegate = mutableMapOf())
+}
```
Comment (Syer10) [inline voyager-core/src/nativeMain/kotlin/cafe.adriel.voyager.core/concurrent/ThreadSafeMap.native.kt:8]:
```suggestion
package cafe.adriel.voyager.core.concurrent

import kotlinx.atomicfu.locks.SynchronizedObject
import kotlinx.atomicfu.locks.synchronized

public actual class ThreadSafeMap<K, V>(
    private val delegate: MutableMap<K, V>
) : MutableMap<K, V> {
    public actual constructor() : this(delegate = mutableMapOf())

    private val syncObject = SynchronizedObject()

    override val size: Int
        get() = synchronized(syncObject) { delegate.size }

    override fun containsKey(key: K): Boolean {
        return synchronized(syncObject) { delegate.containsKey(key) }
    }

    override fun containsValue(value: V): Boolean {
        return synchronized(syncObject) { delegate.containsValue(value) }
    }

    override fun get(key: K): V? {
        return synchronized(syncObject) { delegate[key] }
    }

    override fun isEmpty(): Boolean {
        return synchronized(syncObject) { delegate.isEmpty() }
    }

    override val entries: MutableSet<MutableMap.MutableEntry<K, V>>
        get() = delegate.entries
    override val keys: MutableSet<K>
        get() = delegate.keys
    override val values: MutableCollection<V>
        get() = delegate.values

    override fun clear() {
        synchronized(syncObject) { delegate.clear() }
    }

    override fun put(key: K, value: V): V? {
        return synchronized(syncObject) { delegate.put(key, value) }
    }

    override fun putAll(from: Map<out K, V>) {
        synchronized(syncObject) { delegate.putAll(from) }
    }

    override fun remove(key: K): V? {
        return synchronized(syncObject) { delegate.remove(key) }
    }
}

```

**Review by Syer10 — COMMENTED** [review]
Some suggestions.

Also reguarding the TODOs.
I dont think that you need to be worried about disabling iOS and mac targets on non-apple systems since the kotlin multiplatform plugin ignores them if you are on windows/linux.
For the CI you can probably switch the runner to a MacOS runner since Mac has bash  and usually scripts just work(though I haven't checked the Voyager workflows)

**Comment by zeromod** [comment]
This looks very interesting, it would be really nice to test it in an alpha version. Voyager seems easier to get started compare to decompose.

**Comment by Syer10** [comment]
I have this implemented in my fork of Voyager, I am planning to PR it when my other PRs are accepted https://github.com/Syer10/voyager

**Comment by DevSrSouza** [comment]
Closed in favor of: https://github.com/adrielcafe/voyager/pull/114

**Pull request closed without merging**